### PR TITLE
Random Access Benchmark

### PR DESCRIPTION
**Pull request by eddyxu**
Bring back random access benchmark from POC.

### CODE REVIEW

**Comment by eddyxu** [comment]
Parquet:

```
$ ./point_query --uri s3://eto-ops-testing/pet.parquet
Randomness seeded to: 328404580
Open parquet file: s3://eto-ops-testing/pet.parquet
Number of row groups: 8

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
point_query is a Catch2 v3.0.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Random Access Over One File
-------------------------------------------------------------------------------
/home/ubuntu/work/lance/cpp/benchmarks/point_query.cc:79
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Single Thread                                  100             1     7.98493 m
                                         2.02583 s     1.86021 s     2.21814 s
                                        910.704 ms    733.862 ms     1.13743 s

```

Lance

```
 $ ./point_query --uri s3://eto-ops-testing/pet.lance
Randomness seeded to: 1208603663
Open Lance File: s3://eto-ops-testing/pet.lance
Number of Pages: 8, Rows=7349

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
point_query is a Catch2 v3.0.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Random Access Over One File
-------------------------------------------------------------------------------
/home/ubuntu/work/lance/cpp/benchmarks/point_query.cc:79
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Single Thread                                  100             1     5.89102 m
                                        106.033 ms     96.445 ms    118.931 ms
                                        56.3746 ms    43.7272 ms    71.1296 ms


===============================================================================
All tests passed (102 assertions in 1 test case)
```

File: cpp/benchmarks/point_query.cc [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    std::uniform_int_distribution<int64_t> row_dist(0, num_rows);
+    auto row = tbl->Slice(row_dist(mt), 1);
+  };
+
+  BENCHMARK("Single Thread") { return read(); };
+}
+
+void BenchmarkPointQueryLance(const std::string& uri) {
+  auto f = OpenUri(uri);
+  auto reader = ::lance::arrow::FileReader::Make(f).ValueOrDie();
```
Comment (changhiskhan) [inline cpp/benchmarks/point_query.cc:63]: just making sure this doesn't suffer from the similar issue as the other benchmark where it's loading the whole thing into memory first?

**Review by changhiskhan — APPROVED** [review]
I think if we run this on a nested column we should see much better results right? Nested columns are such a big part of deep learning datasets that it deserves to be treated as a first class citizen here i think

File: cpp/benchmarks/point_query.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/benchmarks/point_query.cc:63]: It should not. The bug happens when we create a Scanner over the FileReader.
here `FileReader::Make(f)` only reads enough FileMetadata.

**Comment by eddyxu** [comment]
> I think if we run this on a nested column we should see much better results right? Nested columns are such a big part of deep learning datasets that it deserves to be treated as a first class citizen here i think

Agreed. Lets add column selection over this PR, like `./point_query --columns id,annotation.label,image --uri s3://...` to enhance the benchmarks.

**Pull request merged**